_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+          uri: lb://car-service
+          predicates:
+            - Path=/home/**
+```
+
+At this point, I could access the car service directly at `http://localhost:8090/cars` and through the gateway at `http://localhost:8080/home`.
+
+## Secure Spring Boot Microservices with OAuth 2.0 and OIDC
+
+To configure the Okta Spring Boot starter, there are a few properties in the `api-gateway` project's `application.properties` file.
+
+```properties
+okta.oauth2.issuer=${OKTA_OAUTH2_ISSUER}
+okta.oauth2.client-id=${OKTA_OAUTH2_CLIENT_ID}
+okta.oauth2.client-secret=${OKTA_OAUTH2_CLIENT_SECRET}
+okta.oauth2.audience=${OKTA_OAUTH2_AUDIENCE}
+```
+
+The `car-service` is configured as an OAuth resource server and has the following properties in its `application.properties` file.
+
+```properties
+okta.oauth2.issuer=${OKTA_OAUTH2_ISSUER}
+okta.oauth2.audience=${OKTA_OAUTH2_AUDIENCE}
+```
+
+The variables are read from the `.env` file in each project's root directory.
+
+### Fetch an Access Token as a JWT
+
+When I first got things working, I was able to log in to the gateway, but when I tried to connect to the downstream microservice, it said the JWT was invalid. For this reason, I added a `/print-token` endpoint to the gateway that prints the access token to the console.
+
+```java
+package com.example.apigateway.web;
+
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.annotation.RegisteredOAuth2AuthorizedClient;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+class HomeController {
+
+    @GetMapping("/")
+    public String howdy(@AuthenticationPrincipal OidcUser user) {
+        return "Hello, " + user.getFullName();
+    }
+
+    @GetMapping("/print-token")
+    public String printAccessToken(@RegisteredOAuth2AuthorizedClient("okta")
+                                   OAuth2AuthorizedClient authorizedClient) {
+
+        var accessToken = authorizedClient.getAccessToken();
+
+        System.out.println("Access Token Value: " + accessToken.getTokenValue());
+        System.out.println("Token Type: " + accessToken.getTokenType().getValue());
+        System.out.println("Expires At: " + accessToken.getExpiresAt());
+
+        return "Access token printed";
+    }
+}
+```
+
+Using jwt.io, I verified that it wasn't a valid JWT. I thought about trying to implement Spring Security's [opaque token support](http://developer.okta.com/blog/2020/08/07/spring-boot-remote-vs-local-tokens), but discovered Auth0 [doesn't have an `/instropection` endpoint](https://community.auth0.com/t/introspection-endpoint-for-opaque-tokens-or-more-flexible-rules-to-get-clear-jwt-access-token/63866). This makes it impossible to use opaque tokens with Auth0.
```

Comment:
may be we should get this clarified with someone. I'm hoping there would be some way to use opaque tokens

Reply:
I'll send you a link to a discussion on Slack.